### PR TITLE
Yield run requests from run status sensors

### PR DESCRIPTION
**Pull request by jamiedemaria**
### Summary & Motivation
ability to yield run requests from run status sensors 

### How I Tested These Changes
bk unit tests and manually checking in dagit

### CODE REVIEW

**Review by sryza — COMMENTED** [review]
Looks great. Only question: should we avoid emitting a `PipelineRunReaction` in the cases where we're firing off a `RunRequest`? What exactly is the meaning of a run reaction?

cc @yuhan

File: python_modules/dagster/dagster/core/definitions/run_status_sensor_definition.py [hunk 1]
```diff
@@ -227,6 +230,14 @@ def pipeline_failure_sensor(
...
         default_status (DefaultSensorStatus): Whether the sensor starts as running or not. The default
             status can be overridden from Dagit or via the GraphQL API.
+        run_request_job (Optional[Union[GraphDefinition, JobDefinition]]): The job a RunRequest should
+            execute if yielded from the sensor.
+        run_request_jobs (Optional[Sequence[Union[GraphDefinition, JobDefinition]]]): (experimental)
+            A list of jobs to be executed if RunRequests are yielded from the sensor.
+        ignore_run_request_job_status (bool): If set to True, the sensor will not execute for any
+            job in run_request_jobs or run_request_job. This prevents an infinite loop of runs created
+            when a run status sensor starts a run and then is triggered again based on the status of that
+            run.
```
[the remark this answers is not in the document]
Comment (jamiedemaria) [reply python_modules/dagster/dagster/core/definitions/run_status_sensor_definition.py:240]: `ignore_run_request_job_status` came about after i tried the following scenario 
* have a run status sensor that triggers on any job success and runs a job called `report_status`
* run a job that succeeds
* the sensor runs `report_status
* `report_status` succeeds
* sensor runs `report_status` again based on the previous `report_status` success
* and on and on with no end 

another option would be to have a negative of the `job_selection` parameter and the sensor will never trigger for any jobs in that list

[the remark this answers is not in the document]
Comment (jamiedemaria) [reply python_modules/dagster/dagster/core/definitions/run_status_sensor_definition.py:240]: also thinking about different names for `run_request_job` that would be more clear, if you have any ideas lmk

**Comment by jamiedemaria** [comment]
@yuhan do you think we can we skip omitting PipelineRunReaction in cases when the sensor doesn't yield a RunRequest (eg in the case where a run status sensor sends an alert to slack)?

File: python_modules/dagster/dagster/core/definitions/run_status_sensor_definition.py [hunk 2]
```diff
@@ -589,9 +624,15 @@ def run_status_sensor(
         description (Optional[str]): A human-readable description of the sensor.
         job_selection (Optional[List[Union[PipelineDefinition, GraphDefinition]]]): Jobs that will
             be monitored by this sensor. Defaults to None, which means the alert will be sent when
-            any job in the repository fails.
+            any job in the repository matches the requested pipeline_run_status.
+        ignore_job_selection (Optional[List[Union[PipelineDefinition, GraphDefinition]]]): Jobs that will
```
[the remark this answers is not in the document]
Comment (jamiedemaria) [reply python_modules/dagster/dagster/core/definitions/run_status_sensor_definition.py:628]: something like this?
```
@run_status_sensor(pipeline_status=DagsterRunStatus.SUCCESS, run_request_job=record_status)
def my_success_sensor(context):
    if context.dagster_run.pipeline_name != "record_status"
         yield RunRequest(....)
    else:
         yield SkipReason(....)
```

File: docs/content/concepts/partitions-schedules-sensors/sensors.mdx [hunk 3]
```diff
@@ -624,6 +624,37 @@ def uses_db_connection():
 
 If a resource you want to initialize has dependencies on other resources, those can be included in the dictionary passed to <PyObject object="build_resources"/>. For more in-depth usage, check out the [Initializing Resources Outside of Execution](/concepts/resources#initializing-resources-outside-of-execution) section.
 
+### Yielding RunRequests in run status sensors
+
+You can also yield <PyObject object="RunRequest" />s from run status sensors.
```
[the remark this answers is not in the document]
Comment (jamiedemaria) [reply docs/content/concepts/partitions-schedules-sensors/sensors.mdx:629]: i like `watched_jobs` or `monitored_jobs` for what we're watching and `response_jobs` for the job we kick off

File: python_modules/dagster/dagster/core/definitions/run_status_sensor_definition.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (prha) [reply python_modules/dagster/dagster/core/definitions/run_status_sensor_definition.py:628]: Yeah, exactly...  you could even implement some sort of retry count using conditionals...

```
retry_count = context.dagster_run.tags.get('my_retry_count', 0)
if retry_count < 3:
    retry_tags = {**context.dagster_run.tags, "my_retry_count": retry_count + 1}
    return RunRequest(tags=retry_tags, ...)
else:
    return SkipReason("can't do it anymoreeee")
```

[the remark this answers is not in the document]
Comment (jamiedemaria) [reply python_modules/dagster/dagster/core/definitions/run_status_sensor_definition.py:628]: i'm good w that

File: docs/content/concepts/partitions-schedules-sensors/sensors.mdx (hunk 3 above)
[the remark this answers is not in the document]
Comment (yuhan) [reply docs/content/concepts/partitions-schedules-sensors/sensors.mdx:629]: watched/monitored vs response/target sounds great.

**Review by prha — DISMISSED** [review]

File: python_modules/dagster/dagster/core/definitions/sensor_definition.py [hunk 4]
```diff
@@ -141,14 +141,18 @@ def repository_name(self) -> Optional[str]:
 SensorExecutionContext = SensorEvaluationContext
 
 RawSensorEvaluationFunctionReturn = Union[
-    Iterator[Union[SkipReason, RunRequest]], List[RunRequest], SkipReason, RunRequest
+    Iterator[Union[SkipReason, RunRequest, PipelineRunReaction]],
```
[the remark this answers is not in the document]
Comment (jamiedemaria) [reply python_modules/dagster/dagster/core/definitions/sensor_definition.py:144]: ok - this is from a mypy complaint because there is a list made later that gets the PipelineRunReactions that are returned by the sensor (if any). other option if to just ignore the mypy warning for that line

File: docs/content/concepts/partitions-schedules-sensors/sensors.mdx [hunk 5]
```diff
@@ -624,6 +624,44 @@ def uses_db_connection():
...
```
[the remark this answers is not in the document]
Comment (jamiedemaria) [reply docs/content/concepts/partitions-schedules-sensors/sensors.mdx]: done!

File: docs/content/concepts/partitions-schedules-sensors/sensors.mdx [hunk 6]
```diff
@@ -624,6 +624,44 @@ def uses_db_connection():
 
 If a resource you want to initialize has dependencies on other resources, those can be included in the dictionary passed to <PyObject object="build_resources"/>. For more in-depth usage, check out the [Initializing Resources Outside of Execution](/concepts/resources#initializing-resources-outside-of-execution) section.
 
+### Yielding RunRequests in run status sensors
```
[the remark this answers is not in the document]
Comment (jamiedemaria) [reply docs/content/concepts/partitions-schedules-sensors/sensors.mdx]: yeah i like this organization a lot better

File: docs/content/concepts/partitions-schedules-sensors/sensors.mdx [hunk 7]
```diff
@@ -402,7 +426,9 @@ def my_slack_on_run_success(context: RunStatusSensorContext):
     )
 ```
 
-Then, you can add the sensor to a repository so it can be enabled and used the same as other sensors:
+As you can see, run status sensors don't have to launch runs of jobs.
```
[the remark this answers is not in the document]
Comment (jamiedemaria) [reply docs/content/concepts/partitions-schedules-sensors/sensors.mdx]: ran a normal sensor and confirmed it get's added to the run event log. updated the page to mention it

File: docs/content/concepts/partitions-schedules-sensors/sensors.mdx [hunk 8]
```diff
@@ -402,7 +427,9 @@ def my_slack_on_run_success(context: RunStatusSensorContext):
     )
 ```
 
-Then, you can add the sensor to a repository so it can be enabled and used the same as other sensors:
+When you write a run status sensor that doesn't return anything, an event is added to the run's event log to indicate that the sensor ran.
```
Comment (yuhan) [inline docs/content/concepts/partitions-schedules-sensors/sensors.mdx:430]:
```suggestion
When a run status sensor is triggered by a job run but doesn't return anything, Dagster will report an event back to the run to indicate that the sensor ran.
```

as im writing this out and per discussion with @prha offline, i find the behavior of run status sensor could expand to be confusing / hard to explain, and i'm now more convinced that we may eventually want to separate 1) run status sensor that acts on run status to start another run, vs 2) a job-level hook that alerts external services on failure/success. 

i think this PR pushes us towards 1) a lot more, while this my_slack_on_run_success is the 2) case. the notes and examples here should be good to address the confusion now, but eventually i feel we may want to move 2) to a separate surface area (other than sensor).

**Review by sryza — COMMENTED** [review]
I think the ship has sailed on the `job` param for sensor. Changing that would affect probably >50% of our users, and I don't think we can make those kinds of changes at this stage in our project's maturity.

Otherwise, this looks great to me.

**Comment by prha** [comment]
@sryza Right, but should we support a list of target jobs for this new `run_status_sensor` parameter?

**Comment by sryza** [comment]
@prha 
> Right, but should we support a list of target jobs for this new run_status_sensor parameter?

Yeah good question. It seems like there are two arguments we could make, right?
- Having both `job` and `jobs` was a mistake, and we shouldn't let that mistake again.
- Having both `job` and `jobs` was a mistake, but it's important to be consistent.

Neither of these seem clearly right-er to me, so I would defer to anyone who has a stronger opinion. If it would be helpful for me to break a tie, I'd say omit `request_job`.

**Review by prha — APPROVED** [review]
I think this looks good.

If we're comfortable just going with `monitored_job` / `request_job` (singular only), I'm good with changing that (cc @sryza ).   But `request_jobs` (multi) is also fine with me... I don't have strong feelings.

**Comment by prha** [comment]
I view it more like this:

* `job` is o.g. param.  `jobs` is experimental param.
* If we had high confidence that `jobs` was the way to go, we'd unexperimentalize it and encourage that instead of `job` at least in documentation.
* If we had high confidence that `jobs` was the way to go, we'd have a single `request_jobs` param to run_status_sensors.

**Comment by sryza** [comment]
> I view it more like this:

Ok yeah I just re-read your comment and realized I misunderstood it the first time.

I don't have high confidence that `jobs` was the way to go.

It seems like the worst possible outcome is that we end up with both `request_job` and `request_jobs` and end up recommending only using the latter.  While this isn't a great outcome, IMO it's not soo bad, given that we'd already be in that situation with regular sensors.

TLDR I would defer to anyone with an opinion.
